File: Lucky Numbers in a Matrix.py
class Solution:
    def luckyNumbers (self, matrix: List[List[int]]) -> List[int]:
        n = len(matrix)
        m = len(matrix[0])
        rows = [float('inf')] * n
        cols = [0] * m
        for i in range(n):
            for j in range(m):
                rows[i] = min(rows[i], matrix[i][j])
                cols[j] = max(cols[j], matrix[i][j])

        cols = set(cols)
        ans = set()
        for i in range(n):
            if rows[i] in cols:
                ans.add(rows[i])
        return ans

File: Lucky Numbers in a Matrix.cpp
class Solution {
public:
    vector<int> luckyNumbers (vector<vector<int>>& matrix) {
        int n = matrix.size(), m = matrix[0].size();
        vector<int> rows(n, INT_MAX);
        vector<int> cols(m, 0);
        for (int i = 0; i< n; i++) {
            for (int j = 0; j < m; j++) {
                rows[i] = min(rows[i], matrix[i][j]);
                cols[j] = max(cols[j], matrix[i][j]);
            }
        }

        unordered_set<int> cols_set(cols.begin(), cols.end());
        vector<int> ans;
        for (int i = 0; i < n; i++) {
            if (cols_set.contains(rows[i])) ans.push_back(rows[i]);
        }
        return ans;
    }
};
